power_grid_model_c/power_grid_model/include/power_grid_model/grouped_index_vector.hpp

```diff
@@ -0,0 +1,255 @@
+                                                                         { t.size() } -> std::same_as<Idx>;
+
+                                                                         { t.begin() } -> index_range_iterator;
+                                                                         { t.end() } -> index_range_iterator;
+                                                                         {
+                                                                             t.get_element_range(idx)
+                                                                             } -> random_access_iterable_like<Idx>;
+
+                                                                         { t.element_size() } -> std::same_as<Idx>;
+                                                                         { t.get_group(idx) } -> std::same_as<Idx>;
+                                                                     };
+} // namespace detail
+
+template <typename T>
+concept grouped_idx_vector_type = detail::grouped_index_vector_type<T>;
+
+struct from_sparse_t {};
+struct from_dense_t {};
+
+constexpr auto from_sparse = from_sparse_t{};
+constexpr auto from_dense = from_dense_t{};
+
+class DenseGroupedIdxVector;
+
+class SparseGroupedIdxVector {
+  private:
+    class GroupIterator : public boost::iterator_facade<GroupIterator, Idx, boost::random_access_traversal_tag,
+                                                        boost::iterator_range<IdxCount>, Idx> {
+      public:
+        using iterator = boost::iterator_range<IdxCount>;
+
+        explicit constexpr GroupIterator(IdxVector const& indptr, Idx group) : indptr_(indptr), group_(group) {}
+
+      private:
+        IdxVector const& indptr_;
+        Idx group_;
+        friend class boost::iterator_core_access;
+
+        iterator dereference() const { return boost::counting_range(indptr_[group_], indptr_[group_ + 1]); }
+        constexpr bool equal(GroupIterator const& other) const { return group_ == other.group_; }
+        constexpr void increment() { ++group_; }
+        constexpr void decrement() { --group_; }
+        constexpr void advance(Idx n) { group_ += n; }
+        constexpr Idx distance_to(GroupIterator const& other) const { return other.group_ - group_; }
+    };
+
+    constexpr auto group_iterator(Idx group) const { return GroupIterator{indptr_, group}; }
+
+  public:
+    auto size() const { return static_cast<Idx>(indptr_.size()) - 1; }
+    auto begin() const { return group_iterator(0); }
+    auto end() const { return group_iterator(size()); }
+    auto get_element_range(Idx group) const { return *group_iterator(group); }
+
+    auto element_size() const { return indptr_.back(); }
+    auto get_group(Idx element) const -> Idx {
+        assert(element < element_size());
+        return std::distance(std::begin(indptr_), std::upper_bound(std::begin(indptr_), std::end(indptr_), element)) -
```

Comment:
what's the reason to use `upper_bound` everywhere instead of `lower_bound`? I guess using `lower_bound` you don't need to use the `-1`?

Reply:
the problem is when there is are repeating values (no elements in a group).

e.g.:

the indptr `{0, 0, 2}` will result in `get_group(0) == 0` instead of the correct value of `get_group(0) == 1`